File: java/phone-number/src/main/java/PhoneNumber.java
class PhoneNumber {
    private final String number;

    PhoneNumber(String numberString) {
        // Remove all non-digit characters
        String digitsOnly = numberString.replaceAll("[^0-9]", "");
        
        // Check for letters
        if (numberString.matches(".*[a-zA-Z].*")) {
            throw new IllegalArgumentException("letters not permitted");
        }
        
        // Check for punctuations (excluding spaces, parentheses, hyphens, dots, and plus sign)
        if (numberString.replaceAll("[0-9\\s()\\-\\.\\+]", "").length() > 0) {
            throw new IllegalArgumentException("punctuations not permitted");
        }
        
        // Validate the length
        if (digitsOnly.length() < 10) {
            throw new IllegalArgumentException("must not be fewer than 10 digits");
        } else if (digitsOnly.length() > 11) {
            throw new IllegalArgumentException("must not be greater than 11 digits");
        } else if (digitsOnly.length() == 11 && digitsOnly.charAt(0) != '1') {
            throw new IllegalArgumentException("11 digits must start with 1");
        }
        
        // Extract the 10-digit number (removing country code if present)
        String tenDigits = digitsOnly.length() == 11 ? digitsOnly.substring(1) : digitsOnly;
        
        // Validate area code
        if (tenDigits.charAt(0) == '0') {
            throw new IllegalArgumentException("area code cannot start with zero");
        } else if (tenDigits.charAt(0) == '1') {
            throw new IllegalArgumentException("area code cannot start with one");
        }
        
        // Validate exchange code
        if (tenDigits.charAt(3) == '0') {
            throw new IllegalArgumentException("exchange code cannot start with zero");
        } else if (tenDigits.charAt(3) == '1') {
            throw new IllegalArgumentException("exchange code cannot start with one");
        }
        
        this.number = tenDigits;
    }

    String getNumber() {
        return number;
    }
}

File: javascript/phone-number/phone-number.js
//
// This is only a SKELETON file for the 'Phone Number' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const clean = (phoneNumber) => {
  // Remove all non-alphanumeric characters
  const digitsOnly = phoneNumber.replace(/\D/g, '');
  
  // Check for letters in the original input (after removing valid punctuation)
  if (/[a-zA-Z]/.test(phoneNumber)) {
    throw new Error('Letters not permitted');
  }
  
  // Check for invalid punctuation (anything that's not a digit, space, parenthesis, dash, dot, or plus)
  if (/[^\d\s()\-\.+]/.test(phoneNumber)) {
    throw new Error('Punctuations not permitted');
  }
  
  // Check the length of the digits
  if (digitsOnly.length < 10) {
    throw new Error('Incorrect number of digits');
  } else if (digitsOnly.length > 11) {
    throw new Error('More than 11 digits');
  } else if (digitsOnly.length === 11 && digitsOnly[0] !== '1') {
    throw new Error('11 digits must start with 1');
  }
  
  // Get the 10-digit number (remove country code if present)
  const tenDigits = digitsOnly.length === 11 ? digitsOnly.substring(1) : digitsOnly;
  
  // Check area code (first 3 digits)
  if (tenDigits[0] === '0') {
    throw new Error('Area code cannot start with zero');
  } else if (tenDigits[0] === '1') {
    throw new Error('Area code cannot start with one');
  }
  
  // Check exchange code (next 3 digits)
  if (tenDigits[3] === '0') {
    throw new Error('Exchange code cannot start with zero');
  } else if (tenDigits[3] === '1') {
    throw new Error('Exchange code cannot start with one');
  }
  
  return tenDigits;
};
